### ISSUE TITLE
overhaul CLI to be more like git

### ISSUE THREAD

**Issue by lukemarsden** [dotmesh-io/dotmesh#251]
while writing the collaboration tutorial, I had to write down the steps that someone has to go through to get a specific version of code and data (kindof the whole point of dotmesh).

it looks like this:

The first thing you need to do is to get the right version of the code.
Looking at the ticket above, you can see that you need to do that with:

```
git clone https://github.com/dotmesh-io/moby-counter
cd moby-counter
git checkout multiple-services
```

Now, before starting up the app, clone the datadot as well:

```
dm clone hub mobycounter_app
dm switch mobycounter_app
```

Now that we've cloned the datadot, we can start up the app:

```
docker-compose up -d
```

And then finally pull down and switch to the specific branch:

```
dm pull hub mobycounter_app security-vulnerability
dm checkout multiple-services
```

---

It's WAY too complicated and sensitive to ordering.

---

This would be better:

Let's get the data in place before we start the app:

```
dm clone user@dothub.com:namespace/mobycounter_app # <-- should pull down all branches
dm switch mobycounter_app
dm checkout multiple-services
```

Then start the app on the data:

```
git clone git@github.com:dotmesh-io/moby-counter
cd moby-counter
git checkout multiple-services
```

Now start the app and the data together!

```
docker-compose up -d
```

(It shouldn't matter which order we do the app and the data clones in. If you
try and start the app before starting the data, it should error. cloning from a
remote is probably more common workflow than creating a new one. That can be
what `dm init` is for.)

Later...

```
dm push
```

... on its own, should work, as the remote has already been set up. Remotes
should be per-repo, and there should be a client-side default remote. This is
what `dm remote -v` should show.

**Comment by deitch** [comment]
I agree, too complicated and brittle. I have a few comments:

```
dm clone user@dothub.com:namespace/mobycounter_app # <-- should pull down all branches
dm switch mobycounter_app
dm checkout security-vulnerability
```

How about

```
$ dm clone user@dothub.com:namespace/mobycounter_app my_moby # <-- should pull down all branches, and pulls to given name
 $ # no dm switch
$ dm checkout my_moby/security-vulnerability # <-- explicit that this is making security-vulnerability the effective branch on my_moby
```

Also, back in late November in Bristol, we had discussed the ability to have a single compose that fetches it, something like:

```
volumes:
  my_moby:
    driver: dm
    source: user@dothub.com:namespace/mobycounter_app:security-vulnerability
```

This would eliminate the manual steps entirely, encapsulating the claim right there in the volume definition. This would do the equivalent of `dm clone` and `dm checkout` all at once.

**Comment by lukemarsden** [comment]
I don't like `dm checkout my_moby/security-vulnerability` because it's _less_ like git to specify the "repo" (dot) and the branch name in the `checkout` command :)

Re your latter point, https://github.com/dotmesh-io/dotmesh/issues/105

**Comment by deitch** [comment]
> it's less like git to specify the "repo" (dot) and the branch name in the checkout command :)

True. But `dotmesh` is **not** git, but git-like. Git changes directories to work in different places. We have a client-server model where, without changing directories, I work on multiple ones. I could work on multiple at the same time. The git paradigm doesn't translate well here.

The bigger issues is the assumption that we always can rely on there being an "activated" or "default" one. Even if we store that in a local (to the client) `.config`, it still means only one user can exist on the system. Go do that in automation systems (Jenkins and other CI, cloud, etc.) and the model falls apart.

**Comment by deitch** [comment]
> Re your latter point, #105

Ah thanks. I thought it was somewhere...

**Comment by lukemarsden** [comment]
```
dm checkout --dot my_moby security-vulnerability
```
I can buy this. And `--dot` and `--branch` supported flags for all commands that ever touch the interactive "current" anything.

**Comment by lukemarsden** [comment]
I've split the latter point into https://github.com/dotmesh-io/dotmesh/issues/266

This issue is for implementing the workflow described in the original comment

**Comment by deitch** [comment]
> I can buy this. And --dot and --branch supported flags for all commands that ever touch the interactive "current" anything.

Oh yes, totally good. My main point was about capability, not semantics.

Although, I will admit that there is something appealing about having dothub.com in the reference of every dot, even local, but that is a different discussion entirely. :-)

**Comment by alaric-dotmesh** [comment]
Related: #223
